File: src/operator/instance_norm.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * Copyright (c) 2015 by Contributors
 * \file instance_norm.cc
 * \brief
 * \author Sebastian Bodenstein
*/

#include "./instance_norm-inl.h"

namespace mxnet {
namespace op {

DMLC_REGISTER_PARAMETER(InstanceNormParam);

struct InstanceNormGrad {
  const char *op_name;
  std::vector<nnvm::NodeEntry> operator()(const nnvm::ObjectPtr& n,
                                          const std::vector<nnvm::NodeEntry>& ograds) const {
    std::vector<nnvm::NodeEntry> out_data;
    out_data.reserve(n->num_outputs());
    for (size_t i = 0; i < n->num_outputs(); ++i)
      out_data.emplace_back(n, i, 0);
    std::vector<nnvm::NodeEntry> heads;
    heads.reserve(5);
    heads.emplace_back(ograds.at(instance_norm::kOut));
    heads.emplace_back(out_data.at(instance_norm::kMean));
    heads.emplace_back(out_data.at(instance_norm::kVar));
    heads.emplace_back(n->inputs.at(instance_norm::kData));
    heads.emplace_back(n->inputs.at(instance_norm::kGamma));

    return MakeGradNode(op_name, n, heads, n->attrs.dict);
  }
};


NNVM_REGISTER_OP(InstanceNorm)
.add_alias("_npx_instance_norm")
.describe(R"code(Applies instance normalization to the n-dimensional input array.

This operator takes an n-dimensional input array where (n>2) and normalizes
the input using the following formula:

.. math::

  out = \frac{x - mean[data]}{ \sqrt{Var[data] + \epsilon}} * gamma + beta

This layer is similar to batch normalization layer (`BatchNorm`)
with two differences: first, the normalization is
carried out per example (instance), not over a batch. Second, the
same normalization is applied both at test and train time. This
operation is also known as `contrast normalization`.

If the input data is of shape [batch, channel, spacial_dim1, spacial_dim2, ...],
`gamma` and `beta` parameters must be vectors of shape [channel].

This implementation is based on this paper [1]_

.. [1] Instance Normalization: The Missing Ingredient for Fast Stylization,
   D. Ulyanov, A. Vedaldi, V. Lempitsky, 2016 (arXiv:1607.08022v2).

Examples::

  // Input of shape (2,1,2)
  x = [[[ 1.1,  2.2]],
       [[ 3.3,  4.4]]]

  // gamma parameter of length 1
  gamma = [1.5]

  // beta parameter of length 1
  beta = [0.5]

  // Instance normalization is calculated with the above formula
  InstanceNorm(x,gamma,beta) = [[[-0.997527  ,  1.99752665]],
                                [[-0.99752653,  1.99752724]]]

)code" ADD_FILELINE)
.add_argument("data", "NDArray-or-Symbol",
              "An n-dimensional input array (n > 2) of the form [batch, "
              "channel, spatial_dim1, spatial_dim2, ...].")
.add_argument("gamma", "NDArray-or-Symbol",
              "A vector of length \'channel\', which multiplies the "
              "normalized input.")
.add_argument("beta", "NDArray-or-Symbol",
              "A vector of length \'channel\', which is added to the "
              "product of the normalized input and the weight.")
.add_arguments(InstanceNormParam::__FIELDS__())
.set_num_inputs(3)
.set_num_outputs(3)
.set_attr<nnvm::FListInputNames>("FListInputNames", [](const NodeAttrs& attrs) {
  return std::vector<std::string>{"data", "gamma", "beta"};
})
.set_attr<nnvm::FListOutputNames>("FListOutputNames",
    [](const NodeAttrs& attrs) {
    return std::vector<std::string>{"output"};
})
.set_attr<nnvm::FNumVisibleOutputs>("FNumVisibleOutputs",
  [](const NodeAttrs& attrs) { return 1; })
.set_attr_parser(ParamParser<InstanceNormParam>)
.set_attr<mxnet::FInferShape>("FInferShape", InstanceNormShape)
.set_attr<THasDeterministicOutput>("THasDeterministicOutput", true)
.set_attr<nnvm::FGradient>("FGradient", InstanceNormGrad{"_backward_instance_norm"})
.set_attr<FCompute>("FCompute<cpu>", InstanceNormForward<cpu>);

NNVM_REGISTER_OP(_backward_instance_norm)
.set_num_inputs(5)
.set_num_outputs(3)
.set_attr_parser(ParamParser<InstanceNormParam>)
.set_attr<FResourceRequest>("FResourceRequest",
  [](const NodeAttrs& attrs) {
    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
  })
.set_attr<nnvm::TIsBackward>("TIsBackward", true)
.set_attr<FCompute>("FCompute<cpu>", InstanceNormBackward<cpu>);

}  // namespace op
}  // namespace mxnet

File: csharp-package/src/MxNet/Gluon/NN/BaseLayers/InstanceNorm.cs
/*****************************************************************************
   Copyright 2018 The MxNet.Sharp Authors. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/
using MxNet.Initializers;

namespace MxNet.Gluon.NN
{
    public class InstanceNorm : HybridBlock
    {
        public InstanceNorm(int axis = 1, float epsilon = 1e-5f, bool center = true, bool scale = false,
            string beta_initializer = "zeros", string gamma_initializer = "ones",
            int in_channels = 0) : base()
        {
            Axis = axis;
            Epsilon = epsilon;
            Center = center;
            Scale = scale;
            In_Channels = in_channels;
            this["gamma"] = Params.Get("gamma", scale ? OpGradReq.Write : OpGradReq.Null, new Shape(in_channels),
                init: Initializer.Get(gamma_initializer), allow_deferred_init: true);
            this["beta"] = Params.Get("beta", center ? OpGradReq.Write : OpGradReq.Null, new Shape(in_channels),
                init: Initializer.Get(beta_initializer), allow_deferred_init: true);
        }

        public int Axis { get; }
        public float Epsilon { get; }
        public bool Center { get; }
        public bool Scale { get; }
        public int In_Channels { get; }
        public Parameter Gamma { get; set; }
        public Parameter Beta { get; set; }

        public override NDArrayOrSymbol HybridForward(NDArrayOrSymbol x, NDArrayOrSymbolList args)
        {
            var gamma = args[0];
            var beta = args[1];

            if (Axis == 1)
            {
                if (x.IsNDArray)
                    return nd.InstanceNorm(x.NdX, gamma.NdX, beta.NdX, Epsilon);
                return sym.InstanceNorm(x.SymX, gamma.SymX, beta.SymX, Epsilon, "fwd");
            }

            if (x.IsNDArray)
            {
                var xs = nd.SwapAxis(x.NdX, 1, (uint)Axis);
                return nd.SwapAxis(nd.InstanceNorm(xs, gamma.NdX, beta.NdX, Epsilon), 1, (uint)Axis);
            }
            else
            {
                var xs = sym.SwapAxis(x.SymX, 1,(uint) Axis);
                return sym.SwapAxis(sym.InstanceNorm(xs, gamma.SymX, beta.SymX, Epsilon, "fwd"), 1, (uint)Axis);
            }
        }

        public override string ToString()
        {
            var in_channels = Params["gamma"].Shape[0];
            return $"{GetType().Name}(eps={Epsilon}, axis={Axis}, cneter={Center}, scale={Scale}, in_channels={in_channels})";
        }
    }
}